src/collection.js

```diff
@@ -460,6 +460,8 @@ export default class Collection {
    * @return {Promise}
    */
   importChanges(syncResultObject, changeObject) {
+    // XXX: Ensure all imports are done within a single transaction
+    // Remove this and replace by batch(changes.map(change => batch[action](change)))
     return Promise.all(changeObject.changes.map(change => {
       return this._importChange(change);
     }))
```

Comment:
Also, `_importChange` includes a [`this.get` call](https://github.com/Kinto/kinto.js/pull/180/files#diff-d2dbe7779d59d4dd594e92248caba0a5R434), which also needs to be done within the batch to keep things safe (actually I think that code is not even safe right now?). But you can't just include it as an extra operation in the batch, because then you won't get the result of the `get` operation out.

Maybe an alternative approach could be:

````js
batch(function executor(transaction) {
  // do lots of stuff here,
  // but call transaction.get, transaction.create, etc.
  // instead of this.get, this.create.
});
````
And then `batch` does:
* prepare transaction,
* execute the executor function passed to it
* resolve from the transaction's `onsuccess` etc. handlers
* the transaction object with which the executor function is called has a `.get` function that works